src/rgw/driver/rados/topic_migration.cc

```diff
@@ -0,0 +1,269 @@
+    RGWObjVersionTracker objv;
+    objv.generate_new_write_ver(dpp->get_cct());
+    r = driver->write_topic_v2(topic, exclusive, objv, y, dpp);
+    if (r == -EEXIST) {
+      continue; // ok, someone else already migrated
+    }
+    if (r < 0) {
+      std::string s = "v1 topic migration for " + topic.name + " failed with: " + cpp_strerror(r);
+      rgw_clog_warn(rados, s);
+      return r;
+    }
+  }
+
+  // remove the v1 topics metadata (this destroys the lock too)
+  r = driver->remove_topics(tenant, &topics_objv, y, dpp);
+  if (r == -ECANCELED) {
+    return 0; // ok, someone else already migrated
+  }
+  if (r < 0) {
+    std::string s = "failed to remove migrated v1 topics obj "
+                  + topics_obj.to_str() + " with: " + cpp_strerror(r);
+    rgw_clog_warn(rados, s);
+    return r;
+  }
+  return r;
+}
+
+} // anonymous namespace
+
+int migrate(const DoutPrefixProvider* dpp,
+            rgw::sal::RadosStore* driver,
+            boost::asio::io_context& context,
+            spawn::yield_context yield)
+{
+  auto y = optional_yield{context, yield};
+
+  ldpp_dout(dpp, 1) << "starting v1 topic migration.." << dendl;
+
+  librados::Rados* rados = driver->getRados()->get_rados_handle();
+  const rgw_pool& pool = driver->svc()->zone->get_zone_params().log_pool;
+  librados::IoCtx ioctx;
+  int r = rgw_init_ioctx(dpp, rados, pool, ioctx);
+  if (r < 0) {
+    ldpp_dout(dpp, 1) << "failed to initialize log pool for listing with: "
+        << cpp_strerror(r) << dendl;
+    return r;
+  }
+
+  // loop over all objects with oid prefix "pubsub."
+  auto filter = rgw::AccessListFilterPrefix(rgw::sal::pubsub_oid_prefix);
+  constexpr uint32_t max = 100;
+  std::string marker;
+  bool truncated = false;
+
+  std::vector<std::string> oids;
+  do {
+    oids.clear();
+    r = rgw_list_pool(dpp, ioctx, max, filter, marker, &oids, &truncated);
+    if (r == -ENOENT) {
+      r = 0;
+      break;
+    }
+    if (r < 0) {
+      ldpp_dout(dpp, 1) << "failed to list v1 topic metadata with: "
+          << cpp_strerror(r) << dendl;
+      return r;
+    }
+
+    std::string msg;
+    for (const std::string& oid : oids) {
+      const auto obj = rgw_raw_obj{pool, oid};
+      if (oid.find(rgw::sal::pubsub_bucket_oid_infix) != oid.npos) {
+        ldpp_dout(dpp, 4) << "migrating v1 topics " << oid << dendl;
+
+        r = migrate_notification(dpp, y, driver, obj);
+        if (r < 0) {
+          msg = "migrating v1 topics " + oid + " failed with: " + cpp_strerror(r);
+          rgw_clog_warn(rados, msg);
+        } else {
+          ldpp_dout(dpp, 4) << "migrating v1 topics " + oid + " completed successfully!" << dendl;
+        }
+
+      } else {
+        ldpp_dout(dpp, 4) << "migrating v1 bucket topics " << oid << dendl;
+
+        r = migrate_topics(dpp, y, driver, obj);
+        if (r < 0) {
+          msg = "migrating v1 bucket topics " + oid + " failed with: " + cpp_strerror(r);
+          rgw_clog_warn(rados, msg);
```

Comment:
if migrate_notification() and migrate_topics() are already calling rgw_clog_warn() for their errors, we don't want to call that again here

Reply:
Done!